benches/aead.rs

```diff
@@ -0,0 +1,194 @@
+        Ok(aead::Nonce::assume_unique_for_key(result))
+    }
+}
+
+macro_rules! function_bench_name {
+    ( $benchmark_name:ident, $algorithm:expr, $operation:ident) => {{
+        const FUNC_NAME_BASE: &str = concat!(
+            stringify!($operation),
+            "_",
+            stringify!($benchmark_name),
+            "_",
+            stringify!($algorithm)
+        );
+
+        FUNC_NAME_BASE.replace("&aead::", "")
+    }};
+}
+
+macro_rules! bench {
+    ( $benchmark_name:ident, $algorithm:expr, $chunk_len:expr, $ad:expr ) => {
+        pub(super) fn $benchmark_name(c: &mut criterion::Criterion) {
+            use super::{NonceSequence, NONCE};
+            use criterion::black_box;
+            use ring::{
+                aead::{self, BoundKey},
+                rand::{SecureRandom, SystemRandom},
+            };
+
+            let rng = SystemRandom::new();
+
+            let mut key_bytes = vec![0u8; $algorithm.key_len()];
+            rng.fill(&mut key_bytes).unwrap();
+
+            {
+                let key = aead::UnboundKey::new($algorithm, &key_bytes).unwrap();
+                let mut key = aead::SealingKey::new(key, NonceSequence::new());
+
+                let mut in_out = black_box(vec![0u8; $chunk_len + $algorithm.tag_len()]);
+
+                c.bench_function(
+                    &function_bench_name!($benchmark_name, $algorithm, encrypt),
+                    |b| {
+                        b.iter(|| -> Result<(), ring::error::Unspecified> {
+                            let aad = aead::Aad::from(black_box($ad));
+                            let _tag = key.seal_in_place_separate_tag(aad, &mut in_out)?;
+                            Ok(())
+                        })
+                    },
+                );
+            }
+
+            {
+                let key =
+                    aead::LessSafeKey::new(aead::UnboundKey::new($algorithm, &key_bytes).unwrap());
+
+                let original_ciphertext = {
+                    let nonce = aead::Nonce::assume_unique_for_key(NONCE);
+                    let mut in_out = vec![0u8; $chunk_len + $algorithm.tag_len()];
+                    let aad = aead::Aad::from($ad);
+                    key.seal_in_place_append_tag(nonce, aad, &mut in_out)
+                        .unwrap();
+                    in_out
+                };
+
+                // Intentionally not inside `black_box` since with a random key, the data inside should not
+                // be optimizable. We also want `reset_vec` to be as fast as possible.
+                let mut ciphertext = original_ciphertext.clone();
+
+                // This is extracted so Rust can optimize this reset to a single memcpy.
+                fn reset_vec(original_ciphertext: &[u8], ciphertext: &mut [u8]) {
+                    for (b, c) in &mut ciphertext.iter_mut().zip(original_ciphertext.iter()) {
+                        *b = *c;
+                    }
+                }
```

Comment:
Couldn't we just use `ciphertext.copy_from_slice(original_ciphertext)`?

Reply:
Good point, thanks for pointing that out. I forgot the method existed.